### PR TITLE
obsolete : ledger-tool json-tail & ledger-bridge

### PR DESCRIPTION
**Pull request by sunnygleason**
#### Problem

* to create UIs based on streaming data, we need a convenient way to tail the ledger file
* (or otherwise see the ledger data stream)
* to avoid massive changes to the websocket API while we're still working out what we need, it could be useful to bridge to a data stream service like PubNub

#### Summary of Changes

* extend ledger-tool to have a json-tail subcommand that produces continuous flat file json output
* create a small nodejs app "ledger-bridge" that can bridge the json to the data stream API

Fixes #

N/A

### CODE REVIEW

**Comment by garious** [comment]
@rob-solana, fyi

File: ledger-tool/src/main.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 use clap::{crate_version, App, Arg, SubCommand};
 use solana::bank::Bank;
-use solana::ledger::{read_ledger, verify_ledger};
+use solana::ledger::{read_ledger, tail_ledger, verify_ledger};
```
Comment (mvines) [inline ledger-tool/src/main.rs]: ack, so I'm working on a [patch](https://github.com/solana-labs/solana/pull/2298/files#diff-3125b196bc2e46166465885cd9ad66ad) that'll conflict hard with this -- I'm porting ledger-tool from src/ledger to src/db_ledger and db_ledger won't permit another process (ledger-tool) to open the ledger at the same time as the fullnode has db_ledger open (*I think*).

**Comment by garious** [comment]
@sunnygleason, I merged #2196 so now it's straightforward to add tests for new functionality.

File: ledger-tool/src/main.rs [hunk 2]
```diff
@@ -39,6 +39,7 @@ fn main() {
         )
         .subcommand(SubCommand::with_name("print").about("Print the ledger"))
         .subcommand(SubCommand::with_name("json").about("Print the ledger in JSON format"))
```
Comment (garious) [inline ledger-tool/src/main.rs]: This command that predates your PR is awkward and puts us on a path where a user would expect a `*-tail` option for every future format. Instead, I'd expect there to be a `--format/-f=json` option. If that existed, then your PR would only add a `--tail` option and be used as `ledger-tool print --tail -f json`.  Pinging @rob-solana to see if he agrees.

File: src/ledger.rs [hunk 3]
```diff
@@ -89,9 +91,17 @@ fn entry_at<A: Read + Seek>(file: &mut A, at: u64) -> io::Result<Entry> {
     read_entry(file, len)
 }
 
-fn next_entry<A: Read>(file: &mut A) -> io::Result<Entry> {
+fn next_entry<A: Read>(file: &mut A, cur_pos: &mut usize) -> io::Result<Entry> {
     let len = deserialize_from(file.take(SIZEOF_U64)).map_err(err_bincode_to_io)?;
-    read_entry(file, len)
+    let result: io::Result<Entry> = read_entry(file, len);
```
Comment (garious) [inline src/ledger.rs]: This type annotation shouldn't be needed.

Comment (garious) [reply src/ledger.rs]: Also, the Rust idiom here is the `?` operator. Looks like this code can be reduced to some small variation of:

```rust
let entry = read_entry(file, len)?;
*cur_pos += SIZEOF_U64 + len;
entry
```

File: src/ledger.rs [hunk 4]
```diff
@@ -417,33 +427,100 @@ impl LedgerWriter {
...
+        if !self.tailing {
+            return match next_entry(&mut self.data, &mut self.cur_pos) {
+                Ok(entry) => Some(Ok(entry)),
+                Err(_) => None,
+            };
+        }
+
+        let entry_result;
+
+        loop {
```
Comment (garious) [inline src/ledger.rs]: This doesn't look like a typical implementation of an iterator. Does this make `next` an `O(n)` operation?

File: ledger-bridge/index.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import PubNub from 'pubnub';
```
Comment (garious) [inline ledger-bridge/index.js]: @mvines, do you want JS and associated dependencies in this git repository? Seems like it belongs elsewhere. Maybe in solana-web3.js?

Comment (mvines) [reply ledger-bridge/index.js]: Keeping the js separate for now is my preference, possibly living in the git repo that hosts the  block explorer source.

File: ledger-tool/src/main.rs (hunk 2 above)
Comment (rob-solana) [reply ledger-tool/src/main.rs]: agreed

File: ledger-tool/src/main.rs (hunk 1 above)
Comment (sunnygleason) [reply ledger-tool/src/main.rs]: @mvines / @garious what's the best alternative for real-time streaming a ledger, would it be replicator API? (haven't dug into that part of the code... yet)

File: ledger-bridge/index.js (hunk 5 above)
Comment (sunnygleason) [reply ledger-bridge/index.js]: I like that. So maybe (if replicator API is the future), this becomes a JS bridge to the replicator API that can stream to different sinks PubNub, Redis, etc.?

File: src/ledger.rs (hunk 4 above)
Comment (garious) [reply src/ledger.rs]: How about we explore that landscape on our end? I'll sync up with some folks on Monday and circle back here.

Comment (sunnygleason) [reply src/ledger.rs]: Sounds like a plan -- in any case, I can use this branch for demo purposes for some amount of time while building out the UI. Thank you in advance for your help!

**Comment by sunnygleason** [comment]
Now that I know the code base (& rust) a little better, I've implemented #2313 as a workaround while the ledger/db_ledger code paths transition just to give the team some more space. (UDP stream versus physical ledger tail is proving totally adequate for development in the meantime)

**Comment by sunnygleason** [comment]
closing this pull request, much of this functionality is moving to blockexplorer project

**Pull request closed without merging**